### PR TITLE
Add `DigestService` and `MailchimpService`

### PR DESCRIPTION
**Pull request by seanh**
These are the two services that will make up the core of the LMS part of the email digests feature.
    
Nothing calls these services yet. Future commits will add an admin page and celery tasks that call them.

## Testing

The [fake `HAPI.get_annotations()` method](https://github.com/hypothesis/lms/blob/3834ae16a6241634bbface69d9b0795a4ba56454/lms/services/h_api.py#L74-L113) in this PR returns some annotations with `authority_provided_id`'s that match some groups in one of our Canvas test courses.

You may need to [log in to Canvas](https://hypothesis.instructure.com/login/canvas) as the `eng+canvasstudent@hypothes.is` and launch [an assignment](https://hypothesis.instructure.com/courses/125/assignments/873) to make sure that you've got the instructors and groups in your local DB. But you don't actually need to create any annotations.

Now you need to find the `h_userid` of the `eng+canvasteacher@hypothes.is` user in your LMS DB. Hint: it's `acct:3a022b6c146dfd9df4ea8662178eac@lms.hypothes.is`.

Now you need to call `DigestService.get_instructor_digest()` with `"acct:3a022b6c146dfd9df4ea8662178eac@lms.hypothes.is"` in the `audience` list and with `since` and `until` times that span the time when you created the student annotation above. Then you need pass the resulting `InstructorDigests` list from that call into `MailchimpService.instructor_digests()`. This won't actually send any emails to Mailchimp because you don't have a Mailchimp API key configured locally. Instead you should see `MailchimpService.send_template()` logging what it would have sent to the console:

```python
$ make shell
>>> from datetime import datetime
>>> from lms.services import DigestService, MailchimpService
>>> digests_svc = request.find_service(iface=DigestService)
>>> mailchimp_svc = request.find_service(iface=MailchimpService)
>>> digests = digests_svc.get_instructor_digest(
        audience=["acct:3a022b6c146dfd9df4ea8662178eac@lms.hypothes.is"],
        since=datetime(year=2023, month=2, day=27),
        until=datetime(year=2023, month=2, day=28),
    )
>>> mailchimp_svc.send_instructor_digests(digests)
2023-02-28 15:01:56,311 INFO  [lms.services.mailchimp:69][MainThread] {'template_name': 'instructor_digest', 'message': {'subaccount': 'mailchimp_digests_subaccount', 'from_email': 'mailchimp_digests_email', 'from_name': 'mailchimp_digests_name', 'to': [{'email': 'eng+canvasteacher@hypothes.is', 'name': 'Hypothesis 101 Teacher'}], 'track_opens': True, 'track_clicks': True, 'global_merge_vars': [{'name': 'num_annotations', 'content': 5}, {'name': 'courses', 'content': [{'title': 'Developer Test Course with Sections Enabled', 'num_annotations': 5}]}]}, 'async': True}
```

### CODE REVIEW

File: lms/services/lti_role_service.py [hunk 1]
```diff
@@ -38,6 +38,39 @@ def get_roles(self, role_description: str) -> List[LTIRole]:
...
+
+        If an `application_instances` argument is given (a list of
+        ApplicationInstance's) then only users belonging to those application
+        instances will be returned.
+
+        This may return multiple users with the same `user_id` and `h_userid`
+        but belonging to different ApplicationInstance's. This happens when a
+        single LMS instance has multiple ApplicationInstance's.
+        """
+        query = self._db.query(User)
```
Comment (seanh) [inline lms/services/lti_role_service.py:56]: FIXME: This should use modern SQLAlchemy (`select()`) if possible

File: lms/services/mailchimp.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+                "global_merge_vars": merge_vars,
+            },
+            "async": True,
+        }
+
+        if self.api_key:
+            mailchimp_client = mailchimp_transactional.Client(self.api_key)
+            mailchimp_client.messages.send_template(params)
+        else:
+            print(params)
```
Comment (seanh) [inline lms/services/mailchimp.py:41]: TODO: Use `logging` here so you can see where the message actually comes from

**Review by jon-betts — COMMENTED** [review]
I find this quite hard to follow. This could be helped with some more hints about why certain things in the data structure need to be there.

I'm not quite sure what to suggest, but I have a few general hints:

 * I think we are doing things in loops that might be better to do as a single task after grouping data
 * Separating the data structure representation and the service methods would be good
 * We might want to use fewer objects in general. I think plain data structures from the `HAPI` would probably be better. We have a lot of different user objects for example
 * I don't think we need to deal with LMS groups or users at all as separate entities in the data we return, except to find the answers to questions. It good to stay in the H world as much as possible, as the carnality is correct
 * There some namespace confusion when saying things like courses, groups or users as it's not clear which type we mean. Some more comments or naming clarity here might help with this
 
I think it would be good to identify some sub tasks and extract them out as separate methods instead of doing everything simultaneously.

Some sub-tasks that stood out to me:

 * Given these H-users, add names to them from LMS data
 * Given these H-users, and these H-groups, who are teachers in which?
 * Given these H-groups, add names and descriptions to them from LMS data
 * Arrange these H-groups into coureses using LMS data

_edit_: I made a gist showing how you might process the data in stages doing some of these tasks separately

* https://gist.github.com/jon-betts/acd4dfda9e35105d8d42598da40df1d6

File: tests/unit/lms/services/mailchimp_test.py [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+                instructor,
+                courses={
+                    course.authority_provided_id: CourseDigest(
+                        course=CourseGroupings([course]),
+                        users={User(username=annotator.h_userid): annotations},
+                    )
+                },
+            )
+            for instructor in instructors
+        }
```
Comment (seanh) [inline tests/unit/lms/services/mailchimp_test.py:133]: Oh ... dear

File: tests/unit/lms/services/mailchimp_test.py [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+                                        ],
+                                    },
+                                ],
+                            },
+                            "async": True,
+                        }
+                    ),
+                ]
+            ).only()
+        )
```
Comment (seanh) [inline tests/unit/lms/services/mailchimp_test.py:205]: Oh ... dear.

This should probably mock `MailchimpService.send_template()` and assert about how that is called rather than asserting about the much more complicated calls that're made to the mailchimp client itself.

But the other problem is the complexity of constructing a fake `digests` object (as would be returned by `DigestService`) to pass in to the method.

File: lms/services/digest.py [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+    Grouping,
+    LTIRole,
+    User,
+)
+from lms.services import h_api
+
+
+@dataclass
+class UserRecords:
+    """All models.User rows for a given h_userid."""
```
Comment (seanh) [inline lms/services/digest.py:20]: Really struggling to come up with a good name for this class but the idea is that it's the concept of an `ApplicationInstance`-independent user object that knows about all the `ApplicationInstance`'d "user records" for a given `h_userid`

File: lms/services/digest.py [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+            else:
+                assert grouping.parent.type == Grouping.Type.COURSE
+                course_groupings.append(grouping.parent)
+
+        return cls(course_groupings)
+
+
+@dataclass
+class CourseDigest:
+    course: CourseGroupings
```
[the remark this answers is not in the document]
Comment (seanh) [reply lms/services/digest.py:121]: Indeed, I don't think `MailchimpService` (which is what calls this) actually cares about the underlying list of `models.Course` objects. It only cares about the abstract concept of a "course" and the "title" of that course

File: lms/services/digest.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+    users: Dict[h_api.User, List[h_api.Annotation]] = field(default_factory=dict)
```
[the remark this answers is not in the document]
Comment (seanh) [reply lms/services/digest.py:122]: It's a list of annotations by a user. I think the external code may want to know about the user itself though, e.g. if it wants to display the username. Things the external code needs to know or may need to know in future:

* The total number of annotations in a course (across all users)
* The total number of annotations (across all users and courses)
* The number of users who have annotated in a course
* The total number of users who have annotated (across all courses)
* Details of the users who have annotated (e.g. their usernames)

**Comment by seanh** [comment]
Closing in favor of https://github.com/hypothesis/lms/pull/5138
